### PR TITLE
(feat) function calling incl. unit tests (first draft)

### PR DESCRIPTION
**Pull request by tobitege**
Implement *function calling* with `file_ops` methods as an example.

**End-user friendly description of the problem this fixes or functionality that this introduces**

- [X] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

I'm 100% sure that OH should offer such a feature, but I'm _architecturally_ not quite sure how to combine it best with the current base.
How the files ended up is somewhat how the LLM's suggested it, and I'm pretty sure, there'll be changes needed.
There's an added `test_llm_tools.py` file with 10 tests.

cc @enyst @xingyaoww what do you guys think?

---

https://litellm.vercel.app/docs/completion/function_call

### CODE REVIEW

File: openhands/runtime/plugins/agent_skills/file_ops/llm_tools.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+
+def get_tool_functions() -> dict[str, Any]:
+    """Returns a dictionary mapping function names to their implementations."""
+    tools = FileTools()
+    return {
+        'open_file': tools.open_file,
+        'search_file': tools.search_file,
+        'search_dir': tools.search_dir,
+        'find_file': tools.find_file,
```
Comment (xingyaoww) [inline openhands/runtime/plugins/agent_skills/file_ops/llm_tools.py:176]: I think we also might need scroll_up / down :)

**Review by xingyaoww — COMMENTED** [review]
Love where this is going!! I think there's a lot of simplification we can do to get this in. I don't have a concrete idea yet but i think we can:

- Make our system/user prompt more composable (e.g., a YAML file) where we can only use part of them for tool-calling API and the rest for non-tool call (compatibility mode). We could also activate/deactivate certain functionality (e.g., browsing / IPython) as needed.
- Try to simplify as much as these file ops as possible (e.g., we can combine all "search" tools into one tool, "scroll_up/down" into ONE "scroll_file", one single file "editor").


I plan to work on some of the above in the next couple days. I think for this PR, i think the easiest way now is to convert the function call into `IPythonCellAction` and execute them properly.

1. Setup the infrastructure for function calling (where you did in llm.py)
2. Start define schemas for file_ops -- I think we can probably have a file called `openhands/runtime/plugins/agent_skills/file_ops/schema.py` to copy whatever you have now.
3. For implementation - I think we can have `openhands/runtime/plugins/agent_skills/file_ops/tool_impl.py` that turns each of the function calls defined in the schema, into an `IPythonAction`. For example:

```python
def open_file(path: str) -> IPythonRunCellAction:
    return IPythonRunCellAction(f"open_file({path})")
```
4. In the agent implementation - when we got the tool call response from LLM provider, we can directly convert these tool_calls into OpenHands Actions (based on the implementation we defined in `openhands/runtime/plugins/agent_skills/file_ops/tool_impl.py`

**Comment by enyst** [comment]
I totally agree, I think we should try it!

I was thinking about this. I wonder if it helps the model. I remember in the old days we had quite a bit of trouble with models that weren't respecting the format we asked for - and, models that implement tools must have fixed just that.

I wonder if it helps the model in any way, even if it was capable before. By the way, Deepseek supports it.

**Comment by tobitege** [comment]
Hey guys,
thanks for the feedback! I'll close this one in favor of #4528 since I forgot to push the branch to repo, so you guys can edit stuff.

**Pull request closed without merging**